core/embed/bootloader/main.c

```diff
@@ -414,54 +508,64 @@ int bootloader_main(void) {
   }
 #endif
 
-  // start the bootloader if no or broken firmware found ...
-  if (firmware_present != sectrue) {
-#ifdef TREZOR_EMULATOR
-    // wait a bit so that the empty lock icon is visible
-    // (on a real device, we are waiting for touch init which takes longer)
-    hal_delay(400);
-#endif
-    // ignore stay in bootloader
-    stay_in_bootloader = secfalse;
-    touched = false;
+  ensure(dont_optimize_out_true * (firmware_present == firmware_present_backup),
+         NULL);
+
+  // start the bootloader ...
+  // ... if user touched the screen on start
+  // ... or we have stay_in_bootloader flag to force it
+  // ... or there is no valid firmware
+  if (touched || stay_in_bootloader == sectrue || firmware_present != sectrue) {
+    screen_t screen;
+    if (header_present == sectrue) {
+      ui_set_initial_setup(false);
+      screen = SCREEN_INTRO;
+    } else {
+      screen = SCREEN_WELCOME;
+
+      // erase storage
+      ensure(flash_area_erase_bulk(STORAGE_AREAS, STORAGE_AREAS_COUNT, NULL),
+             NULL);
 
-    ui_set_initial_setup(true);
+      ui_set_initial_setup(true);
 
-    // keep the model screen up for a while
+      // keep the model screen up for a while
 #ifndef USE_BACKLIGHT
-    hal_delay(1500);
+      hal_delay(1500);
 #else
-    // backlight fading takes some time so the explicit delay here is shorter
-    hal_delay(1000);
+      // backlight fading takes some time so the explicit delay here is
+      // shorter
+      hal_delay(1000);
 #endif
-
-    // show welcome screen
-    ui_screen_welcome();
-
-    // erase storage
-    ensure(flash_area_erase_bulk(STORAGE_AREAS, STORAGE_AREAS_COUNT, NULL),
-           NULL);
-
-    // and start the usb loop
-    if (bootloader_usb_loop(NULL, NULL) != CONTINUE) {
-      return 1;
     }
-  }
-
-  // ... or if user touched the screen on start
-  // ... or we have stay_in_bootloader flag to force it
-  if (touched || stay_in_bootloader == sectrue) {
-    ui_set_initial_setup(false);
-
-    screen_t screen = SCREEN_INTRO;
 
     while (true) {
-      bool continue_to_firmware = false;
+      volatile secbool continue_to_firmware = secfalse;
+      volatile secbool continue_to_firmware_backup = secfalse;
       uint32_t ui_result = 0;
 
       switch (screen) {
+        case SCREEN_WELCOME:
+
+          ui_screen_welcome();
+
+          // and start the usb loop
+          switch (bootloader_usb_loop(NULL, NULL)) {
+            case CONTINUE_TO_FIRMWARE:
+              continue_to_firmware = sectrue;
+              continue_to_firmware_backup = sectrue;
+              break;
+            case RETURN_TO_MENU:
+              break;
+            default:
+            case SHUTDOWN:
+              return 1;
+              break;
```

Comment:
we probably want to call `trezor_shutdown()` here too?

Reply:
Returning from `main` will ultimately have the same effect (=shutdown_privileged is called with infinite loop at the end), so i don't think that matters much. Returning from `real_jump_to_firmware` doesn't do this immediately - although it will end up same in this case as there is nothing more to be done in main. Thats until some more calls are added in between and the information is lost, so better to be explicit. Also `return 1` from `void real_jump_to_firmware(void) ` does not compile so the return code is lost - not that its used at this moment.